driver/Src/CAN.c

```diff
@@ -0,0 +1,377 @@
+#define CAN1_SEND_QUEUE_SIZE (10)
+#endif
+
+// can1 handle
+static CAN_HandleTypeDef hcan1_ = {.Instance = CAN1};
+CAN_HandleTypeDef* hcan1 = &hcan1_;
+
+// can1 send semaphore
+static SemaphoreHandle_t can1_send_semaphore = NULL;
+static StaticSemaphore_t can1_send_semaphore_buffer;
+
+// can1 send queue
+static QueueHandle_t can1_send_queue = NULL;
+static StaticQueue_t can1_send_queue_buffer;
+static uint8_t can1_send_queue_storage[CAN1_SEND_QUEUE_SIZE * sizeof(tx_payload_t)];
+
+// can1 recv semaphore
+static SemaphoreHandle_t can1_recv_semaphore = NULL;
+static StaticSemaphore_t can1_recv_semaphore_buffer;
+
+// create recv queue storage
+#define CAN_RECV_ENTRY(ID_, SIZE_) \
+    static uint8_t recv_queue_storage_##ID_[SIZE_ * sizeof(rx_payload_t)]; \
+
+#include "can_entries.h"
+
+#undef CAN_RECV_ENTRY
+
+// create recv queue array
+#define CAN_RECV_ENTRY(ID_, SIZE_) \
+    { \
+        .id = (ID_), \
+        .size = (SIZE_), \
+        .queue = NULL, \
+        .storage = recv_queue_storage_##ID_, \
+        .buffer = {{0}} \
+    },
+
+static recv_entry_t can1_recv_entries[] = {
+    #include "can_entries.h"
+};
+#undef CAN_RECV_ENTRY
+
+// calculate amount of can1 recv entries
+static const uint32_t can1_recv_entry_count = sizeof(can1_recv_entries) / sizeof(can1_recv_entries[0]);
+#endif
+
+// safety check send/recv
+static bool initialized = false;
+
+// CAN MSP init
+void HAL_CAN_MspInit(CAN_HandleTypeDef *hcan) {
+    GPIO_InitTypeDef init = {0};
+
+    if (hcan->Instance == CAN1) {
+        // enable clocks
+        __HAL_RCC_CAN1_CLK_ENABLE();
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+
+        /* enable gpio
+        PA11 -> CAN1_RX
+        PA12 -> CAN1_TX
+        */
+        init.Pin = GPIO_PIN_11;
+        init.Mode = GPIO_MODE_AF_PP;
+        init.Pull = GPIO_PULLUP;
+        init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        init.Alternate = GPIO_AF9_CAN1;
+        HAL_GPIO_Init(GPIOA, &init);
+
+        init.Pin = GPIO_PIN_12;
+        init.Mode = GPIO_MODE_AF_PP;
+        init.Pull = GPIO_NOPULL;
+        init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        init.Alternate = GPIO_AF9_CAN1;
+        HAL_GPIO_Init(GPIOA, &init);
+
+        // enable interrupts
+        HAL_NVIC_SetPriority(CAN1_TX_IRQn, 5, 0);
+        HAL_NVIC_EnableIRQ(CAN1_TX_IRQn);
+        HAL_NVIC_SetPriority(CAN1_RX0_IRQn, 5, 0);
+        HAL_NVIC_EnableIRQ(CAN1_RX0_IRQn);
+    }
+}
+
+// CAN MSP deinit
+void HAL_CAN_MspDeInit(CAN_HandleTypeDef *hcan) {
+    if (hcan->Instance == CAN1) {
+        // disable clocks
+        __HAL_RCC_CAN1_CLK_DISABLE();
+
+        /* disable gpio
+        PA11 -> CAN1_RX
+        PA12 -> CAN1_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_11);
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_12);
+
+        // disable interrupts
+        HAL_NVIC_DisableIRQ(CAN1_TX_IRQn);
+        HAL_NVIC_DisableIRQ(CAN1_RX0_IRQn);
+    }
+}
+
+can_status_t can_init(CAN_HandleTypeDef* handle, CAN_FilterTypeDef* filter) {
```

Comment:
nitpick:
- i'm still gonna move this stuff into a folder called bsp/ later just so the driver folder is reserved for the sd card driver, the seven segment LED driver, any other shared **device** drivers specifically
- BUT i like the naming convention and structure of this driver so i'll probably try to get the other BSP drivers to look more like this